File: core/src/apps/zcash/get_viewing_key.py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from buffer_types import AnyBuffer

    from trezor.messages import ZcashGetViewingKey, ZcashViewingKey


def _call_native(
    seed: bytes,
    network: int,
    account: int,
    output: AnyBuffer,
) -> None:
    from trezorzcash import derive_viewing_key

    derive_viewing_key(seed, network, account, output)


def _derive_viewing_key(
    seed: bytes,
    network: int,
    account: int,
    output: AnyBuffer,
) -> None:
    from trezor import utils

    try:
        _call_native(seed, network, account, output)
    finally:
        # Clear completed native stack frames before any subsequent await.
        utils.zero_unused_stack()


def _derive_seed_fingerprint(seed: bytes, output: AnyBuffer) -> None:
    from trezor import utils
    from trezorzcash import seed_fingerprint

    try:
        seed_fingerprint(seed, output)
    finally:
        utils.zero_unused_stack()


async def get_viewing_key(msg: ZcashGetViewingKey) -> ZcashViewingKey:
    from trezor import TR, utils, wire
    from trezor.enums import ButtonRequestType
    from trezor.messages import ZcashViewingKey
    from trezor.ui.layouts import confirm_action, show_warning
    from trezor.ui.layouts.progress import progress

    from apps.common import coininfo, seed

    from . import helpers, unified_addresses

    if not utils.USE_ZCASH_SHIELDED:
        raise wire.ProcessError("Zcash shielded support is not enabled")

    network = msg.network  # local_cache_attribute
    account = msg.account  # local_cache_attribute
    coin_name, network_label, coin_type = helpers.validate_network_account(
        network, account
    )

    seed.raise_if_not_initialized()
    session = helpers.snapshot_session()
    account_label = helpers.account_label(account)
    path = helpers.account_path(coin_type, account)

    await confirm_action(
        "zcash_export_viewing_key",
        TR.zcash__export_viewing_key,
        action=f"Zcash {network_label}\n{account_label}\n{path}",
        description=TR.zcash__viewing_key_warning,
        br_code=ButtonRequestType.SignTx,
        prompt_screen=True,
        # Hold: an exported viewing key reveals the account's whole history
        # and cannot be revoked.
        hold=True,
    )
    helpers.require_session(session)

    # The viewing key is per account; the ZIP-32 seed fingerprint is not. It
    # is the same value for every account index and for both networks, so it
    # links everything this seed ever signs. The screen above promises account
    # scope, so the fingerprint needs its own opt-in and its own warning.
    include_fingerprint = bool(msg.include_seed_fingerprint)
    if include_fingerprint:
        await show_warning(
            br_name="zcash_seed_fingerprint",
            content=TR.zcash__seed_fingerprint_warning,
            br_code=ButtonRequestType.Warning,
        )
        helpers.require_session(session)

    if helpers.has_weak_backup():
        await show_warning(
            br_name="zcash_weak_backup",
            content=TR.zcash__weak_backup_warning,
            br_code=ButtonRequestType.Warning,
        )
        helpers.require_session(session)

    wallet_seed = await seed.get_seed()
    raw_fvk = bytearray(96)
    # The ZIP-32 seed fingerprint is a public identifier of the seed (a
    # one-way hash), derived only when the request asked for it and the user
    # acknowledged the extra warning above.
    fingerprint = bytearray(32) if include_fingerprint else None
    # Show progress during the blocking native derivation. It is a static
    # ring: `ProgressLayout` only moves on `report()`, and nothing can call it
    # while the native call holds the VM.
    progress_layout = progress(indeterminate=True)
    progress_layout.report(0)
    try:
        helpers.require_session(session)
        try:
            _derive_viewing_key(wallet_seed, network, account, raw_fvk)
            key = unified_addresses.encode_fvk(raw_fvk, coininfo.by_name(coin_name))
            if fingerprint is not None:
                _derive_seed_fingerprint(wallet_seed, fingerprint)
        except (KeyError, ValueError, RuntimeError):
            raise wire.ProcessError("Zcash viewing key derivation failed")
    finally:
        utils.memzero(raw_fvk)
        del wallet_seed

    helpers.require_session(session)
    return ZcashViewingKey(
        key=key,
        seed_fingerprint=None if fingerprint is None else bytes(fingerprint),
    )

File: sdk/apps/zcash/src/get_viewing_key.rs
use ironwood::receive::derive_full_viewing_key_from_spending_key;
use trezor_app_sdk::ui::{self, ConfirmAction};
use trezor_app_sdk::{Error, Result};
use zeroize::Zeroizing;

use crate::account::{
    account_from_request, account_keys, account_label, account_path, network_from_request,
    network_label,
};
use crate::layout::{show_warning, show_weak_backup_warning, with_progress};
use crate::proto::common::button_request::ButtonRequestType;
use crate::proto::zcash::{ZcashGetViewingKey, ZcashViewingKey};
use crate::unified;

/// Exports the account's Orchard-only UFVK, and the seed fingerprint if asked,
/// after the user consents to each.
pub(crate) fn get_viewing_key(msg: ZcashGetViewingKey) -> Result<ZcashViewingKey> {
    let network = network_from_request(msg.network)?;
    let account = account_from_request(msg.account)?;
    let include_seed_fingerprint = msg.include_seed_fingerprint.unwrap_or(false);

    // Hold: an exported viewing key reveals the account's whole history and
    // cannot be revoked.
    let scope = uformat!(
        "Zcash {}\n{}\n{}",
        network_label(network),
        account_label(account).as_str(),
        account_path(network, account).as_str()
    );
    ui::error_if_not_confirmed(ui::confirm_action(ConfirmAction::new(
        tr!("zcash__export_viewing_key"),
        &scope,
        Some(tr!("zcash__viewing_key_warning")),
        None,
        true, // hold
        None,
        false, // cancel
        Some("zcash_export_viewing_key"),
        ButtonRequestType::SignTx.into(),
        false, // external_menu
    ))?)?;

    // The seed fingerprint is the same for every account and both networks,
    // so it links everything this seed signs. The screen above promises
    // account scope, so the fingerprint needs its own opt-in and warning.
    if include_seed_fingerprint {
        show_warning(
            tr!("zcash__seed_fingerprint_warning"),
            "zcash_seed_fingerprint",
        )?;
    }

    let keys = account_keys(network, account)?;
    if keys.weak_backup {
        show_weak_backup_warning()?;
    }
    let mut full_viewing_key = Zeroizing::new([0u8; 96]);
    with_progress(
        |progress| {
            derive_full_viewing_key_from_spending_key(
                &keys.spending_key,
                &mut full_viewing_key,
                progress,
            )
        },
        Error::DataError("Zcash viewing key derivation failed"),
    )?;
    let seed_fingerprint = include_seed_fingerprint.then(|| keys.seed_fingerprint.to_vec());
    drop(keys);

    Ok(ZcashViewingKey {
        key: unified::viewing_key(network, *full_viewing_key)?,
        seed_fingerprint,
    })
}
